test/test_det_noise_sim.cpp

```diff
@@ -68,150 +67,224 @@ TEST(DeterministicNoiseSimTest, TestingBarrierGate) {
+
+    EXPECT_EQ(ddsim->getNumberOfQubits(), 4);
+    EXPECT_EQ(ddsim->getActiveNodeCount(), 22);
+    EXPECT_EQ(ddsim->getMaxNodeCount(), 44);
+    EXPECT_EQ(ddsim->getMaxMatrixNodeCount(), 0);
+    EXPECT_EQ(ddsim->getMatrixActiveNodeCount(), 0);
+    EXPECT_EQ(ddsim->countNodesFromRoot(), 23);
+}
+
+TEST(DeterministicNoiseSimTest, TestFunctionsUnOptimized) {
+    auto quantumComputation = detGetAdder4Circuit();
+    auto ddsim              = std::make_unique<DeterministicNoiseSimulator<>>(quantumComputation, std::string("APD"), 0.01, 0.02, 1, true);
+    auto m                  = ddsim->DeterministicSimulate();
+
+    EXPECT_EQ(ddsim->getNumberOfQubits(), 4);
+    EXPECT_EQ(ddsim->getActiveNodeCount(), 29);
+    EXPECT_EQ(ddsim->getMaxNodeCount(), 58);
+    EXPECT_EQ(ddsim->getMaxMatrixNodeCount(), 0);
+    EXPECT_EQ(ddsim->getMatrixActiveNodeCount(), 0);
+    EXPECT_EQ(ddsim->countNodesFromRoot(), 30);
+
+    EXPECT_TRUE(ddsim->toBinaryString(9, 6) == "100100");
+}
+
+TEST(DeterministicNoiseSimTest, TestingSimulatorFunctionality) {
+    auto quantumComputation = detGetAdder4Circuit();
+    auto ddsim              = std::make_unique<DeterministicNoiseSimulator<>>(quantumComputation, std::string("APD"), 0.01, 0.02, 1, true);
+    auto m                  = ddsim->DeterministicSimulate();
+
+    EXPECT_EQ(ddsim->getNumberOfQubits(), 4);
+    EXPECT_EQ(ddsim->getActiveNodeCount(), 29);
+    EXPECT_EQ(ddsim->getMaxNodeCount(), 58);
+    EXPECT_EQ(ddsim->getMaxMatrixNodeCount(), 0);
+    EXPECT_EQ(ddsim->getMatrixActiveNodeCount(), 0);
+    EXPECT_EQ(ddsim->countNodesFromRoot(), 30);
+}
+
+TEST(DeterministicNoiseSimTest, sampleFromProbabilityMap1) {
+    auto quantumComputation = std::make_unique<qc::QuantumComputation>(2);
+    quantumComputation->emplace_back<qc::StandardOperation>(2, 0, qc::X);
+
+    auto ddsim = std::make_unique<DeterministicNoiseSimulator<>>(quantumComputation, std::string("APD"), 0, 0, 1, true);
+    auto m     = ddsim->DeterministicSimulate();
+
+    std::vector<std::complex<dd::fp>> temp{};
+
+    size_t shots        = 100000;
+    auto   sampledShots = ddsim->sampleFromProbabilityMap(m, shots);
+
+    double tolerance = 0.01;
+    for (auto& result: m) {
+        EXPECT_NEAR(result.second, (dd::fp)sampledShots.find(result.first)->second / shots, tolerance);
+    }
 }
+
+TEST(DeterministicNoiseSimTest, sampleFromProbabilityMap2) {
+    auto quantumComputation = detGetAdder4Circuit();
+
+    auto ddsim = std::make_unique<DeterministicNoiseSimulator<>>(quantumComputation, std::string("AP"), 0.1, 0.2, 1, true);
+
+    std::vector<std::complex<dd::fp>> temp{};
+
+    size_t shots        = 1000000;
+    auto   sampledShots = ddsim->Simulate(shots);
+    auto   m            = ddsim->DeterministicSimulate();
+
+    double tolerance = 0.01;
+    for (auto& result: m) {
+        EXPECT_NEAR(result.second, (dd::fp)sampledShots.find(result.first)->second / shots, tolerance);
+    }
+}
```

Comment:
there is a newline missing at the end of this file

Reply:
Added the newline